x-pack/plugins/security_solution/server/lib/detection_engine/signals/threshold/find_threshold_signals.ts

```diff
@@ -73,75 +103,139 @@ export const findThresholdSignals = async ({
+  do {
+    // Generate a composite aggregation considering each threshold grouping field provided, appending leaf
+    // aggregations to 1) filter out buckets that don't meet the cardinality threshold, if provided, and
+    // 2) return the first and last hit for each bucket in order to eliminate dupes and reconstruct an accurate
+    // timeline.
+    const aggregations: ThresholdAggregationContainer = {
+      thresholdTerms: hasThresholdFields(threshold)
+        ? {
+            composite: {
+              sources: threshold.field.map((term, i) => ({
+                [term]: {
+                  terms: {
+                    field: term,
+                  },
+                },
+              })),
+              after: sortKeys,
+              size: 10000,
+            },
+            aggs: {
+              ...leafAggs,
+              count_check: {
+                bucket_selector: {
+                  buckets_path: {
+                    docCount: '_count',
+                  },
+                  script: `params.docCount >= ${threshold.value}`,
+                },
+              },
             },
-            ...orderByCardinality(),
-            min_doc_count: threshold.value,
+          }
+        : {
+            terms: {
+              script: {
+                source: '""', // Group everything in the same bucket
+                lang: 'painless',
+              },
+              min_doc_count: threshold.value,
+              ...(shouldFilterByCardinality(threshold)
+                ? { order: { cardinality_count: 'desc' } }
+                : {}),
+            },
+            aggs: leafAggs,
           },
-          aggs: leafAggs,
+    };
+
+    const { searchResult, searchDuration, searchErrors } =
+      await singleSearchAfter<ThresholdAggregateContainer>({
+        aggregations,
+        searchAfterSortId: undefined,
+        timestampOverride,
+        index: inputIndexPattern,
+        from,
+        to,
+        services,
+        logger,
+        // @ts-expect-error refactor to pass type explicitly instead of unknown
+        filter,
+        pageSize: 0,
+        sortOrder: 'desc',
+        buildRuleMessage,
+      });
+
+    const thresholdTerms = searchResult.aggregations?.thresholdTerms;
+    const prevThresholdTerms = mergedSearchResults.aggregations?.thresholdTerms;
+
+    sortKeys = isCompositeAggregate(thresholdTerms) ? thresholdTerms.after_key : undefined;
+    mergedSearchResults = mergeSearchResults([mergedSearchResults, searchResult]);
+
+    if (prevThresholdTerms) {
+      const prevBuckets = Array.isArray(prevThresholdTerms.buckets)
+        ? prevThresholdTerms.buckets
+        : []; // TODO: convert object keys to array?
+      const nextBuckets = Array.isArray(thresholdTerms?.buckets)
+        ? thresholdTerms?.buckets ?? []
+        : []; // TODO: convert object keys to array?
+
+      mergedSearchResults.aggregations = {
+        thresholdTerms: {
+          buckets: [...prevBuckets, ...nextBuckets],
```

Comment:
Let's try to move away from the pattern of merging and mutating responses...I think we can just keep an array of bucket arrays here instead of spreading them into a single array? Then we can just push the new buckets onto the array of arrays and not worry about `prevBuckets` at all.